Bug #2179: Wrong owner_type on the request returns 500

dLobatog opened the pull request:
http://projects.theforeman.org/issues/2179

This pull request validates the owner type. This will fix the problem where the API returns 500 with a malformed request that contains a wrong owner_type. Instead it returns a 422 with info on how to solve the error.

dLobatog committed:
Validate the owner type. This will fix the problem where the API returns 500 with a malformed request that contains a wrong owner_type

Validate owner type on user model to return 422

Use validates_inclusion_of to validate owner type in Host::Base

Blank owner type is not a valid value

Use OWNER_TYPES and allow empty owner_type for hosts other than Host::Managed

Added test for empty owner_type when creating unmanaged host (should pass)

ohadlevy commented on app/models/host.rb:
```diff
@@ -213,8 +215,9 @@ class Jail < ::Safemode::Jail
-  after_validation :ensure_associations
+  before_validation :set_hostgroup_defaults, :set_ip_address, :normalize_addresses, :normalize_hostname, :force_lookup_value_matcher
+  after_validation :ensure_associations 
+  before_save :set_default_user
```
can you remind me again why did changed into before_save?

dLobatog replied:
The point of this commit is to check the owner type so that you return the right return type (403 because the owner type is invalid). 
This is done by a validation of the owner_type attribute. 
If owner_type is wrong, 'set_default_user' will fail right away (it uses 'self.owner') and return a 500 which is what we want to avoid, as this is not an internal server error, if we get to this point (validations and such) the request is well formed, but the data inside the request might be wrong. 
Instead of doing this, as 'set_default_user' needs to be saved but it doesn't deal with any of the validations, we should defer 'set_default_user' to any stage after the validation, to ensure it gets the right data (= not a broken owner type). 

Since any failures on the owner_type will be handled during the validation, both 'after_validation' and 'before_save' are proper places to execute this method with a guarantee that it won't fail.

ohadlevy replied:
the thing i'm worried about, is if you would like to know the owner during the orchestration (which happens between validations and save).

maybe change the default_owner to happen only if owner_type and owner_id is set?

dLobatog replied:
That's a good point, it's either that or (if orchestration happens after validations) maybe set_default_user could go in the 'after_validation'. If we leave it in the "before_validation" spot,  "set_default_user" runs before verifying the 'owner type' is correct and sometimes breaks (because we haven't yet done the validation) .

ohadlevy replied:
maybe in this case we should do the validation and set the default one in
one method?

dLobatog replied:
Kind of odd behavior I think because it couples the code but yeah we can do both things on the validation if that's what you want. Will push that tomorrow

dLobatog replied:
I just noticed :ensure_owner_type is only validated if owner_type is present.. so :set_default_user cant go there as it will not be executed sometimes. Maybe 'after_validation' is the right place?

ohadlevy replied:
its been a while so maybe i'm just BS but can't we set the default can happen before validations? (only if the value doesnt exists in the first place)?

dLobatog replied:
http://stackoverflow.com/questions/7606244/how-should-my-rails-before-validation-callbacks-handle-bad-data

I got around this by letting the error slip through on the before_validation call (set_default_user), since it will be catched on the actual validation (ensure_owner_type). Not pretty but apparently a convention according to the SO user

dLobatog commented:
Rebased and fixed after @ohadlevy 's comment about "before_save" calls possibly conflicting with retrieving the owner on orchestration.

isratrade commented:
@eLobato, I'm not following as owner_type is populated by Rails as User and Host have a has_many polymorphic relationship :as => :owner.  I can't imaged a situation where host.owner_id exists but host.owner_type does not exist?

Also, issue #2179 show the parameters are :owner_type => "UserGroup" whereas it should be "Usergroup".

dLobatog commented:
@isratrade You are right, host.owner_type will always exist. The problem is that you can submit requests (I tried with the API, not sure if there are other methods) without that field set/with a wrong value, like "UserGroup"

That made the before_validation set_default_user call crash and return a 500. That's exactly the issue. 

In order to avoid this crash and the HTTP 500 that will be returned to the user, this pull request fixes this and returns a 422 (Unprocessable entity), doesn't crash, and gives an informative message to the user.

isratrade commented:
@eLobato, ok, I see, create host using API needs :owner_type.  So it's an error handling mechanism for faulty user input.

ohadlevy commented on app/models/host/base.rb:
```diff
@@ -13,6 +13,7 @@ class Base < ActiveRecord::Base
     validates_presence_of   :name
     validates_uniqueness_of :name
     validate :is_name_downcased?
+    validate :ensure_owner_type, :if => Proc.new { |host| host.owner_type.present?}
```
can't we use the built in rails validator instead?
e.g.

``` ruby
   validates_inclusion_of :owner_type
                           :in          => TYPES,
                           :allow_blank => true,
                           :message     => (_("Please select owner type from one of the following: %s") % TYPES.join(', '))
```

dLobatog commented:
I have just updated this using a more appropriate validator suggested by @ohadlevy

domcleal commented on app/models/host/base.rb:
```diff
@@ -4,6 +4,7 @@ module Host
   class Base < ActiveRecord::Base
     include Foreman::STI
     self.table_name = :hosts
+    TYPES = %w(User Usergroup)
```
Nitpick, OWNER_TYPES would be clearer within Host.

domcleal commented on app/models/host/managed.rb:
```diff
@@ -793,6 +794,7 @@ def to_ip_address name_or_ip
   end
 
   def set_default_user
+    return unless %w(User Usergroup).include?(self.owner_type)
```
Could you use the TYPES constant instead of repeating?

ohadlevy commented on app/models/host.rb:
```diff
@@ -27,4 +27,5 @@ def self.respond_to?(method, include_private = false)
     end
   end
 
+
```
nitpick, WHITESPACE

dLobatog commented:
Updated the PR fixing the comments above, and I added a test for unmanaged hosts, which should be created even when the owner_type is not supplied. 

Sorry for the '%w(User Usergroup)' all over the place!

domcleal commented:
Thanks @eLobato, merged as 7fcb314 for Foreman 1.3.
